File: app/src/main/java/com/example/ui/common/RouteOptimizer.kt
package com.example.ui.common

data class MapSurveyItem(
    val id: String,
    val isUnverified: Boolean,
    val latitude: Double,
    val longitude: Double,
    val title: String,
    val description: String,
    val price: Double? = null,
    val propertyType: String? = null,
    val status: String? = null,
    val distanceKm: Double? = null,
    val areaSize: Double? = null,
    val imagePath: String? = null,
    val driveMediaIds: String? = null
)

data class OptimizationResult(
    val optimizedPoints: List<MapSurveyItem>,
    val invalidPointsCount: Int
)

object RouteOptimizer {

    fun optimize(
        start: Pair<Double, Double>?,
        points: List<MapSurveyItem>
    ): OptimizationResult {
        // Filter out coordinates that are 0,0 or null (invalid)
        val validPoints = points.filter { MapsIntentHelper.isValidCoordinate(it.latitude, it.longitude) }
        val invalidCount = points.size - validPoints.size

        if (validPoints.isEmpty()) {
            return OptimizationResult(emptyList(), invalidCount)
        }

        val optimized = if (validPoints.size <= 9) {
            permute(start, validPoints)
        } else {
            twoOpt(start, validPoints)
        }

        return OptimizationResult(optimized, invalidCount)
    }

    private fun permute(
        start: Pair<Double, Double>?,
        points: List<MapSurveyItem>
    ): List<MapSurveyItem> {
        if (points.isEmpty()) return points
        
        var bestRoute = points
        var minDistance = Double.MAX_VALUE
        
        val indices = points.indices.toList()
        val permutations = mutableListOf<List<Int>>()
        generatePermutations(indices, 0, permutations)
        
        for (perm in permutations) {
            var currentDist = 0.0
            var prevLat = start?.first
            var prevLng = start?.second
            
            for (idx in perm) {
                val pt = points[idx]
                if (prevLat != null && prevLng != null) {
                    currentDist += com.example.util.GeoUtils.haversineKm(
                        prevLat, prevLng, pt.latitude, pt.longitude
                    )
                }
                prevLat = pt.latitude
                prevLng = pt.longitude
            }
            
            if (currentDist < minDistance) {
                minDistance = currentDist
                bestRoute = perm.map { points[it] }
            }
        }
        
        return bestRoute
    }

    private fun generatePermutations(list: List<Int>, k: Int, result: MutableList<List<Int>>) {
        val n = list.size
        val mutableList = list.toMutableList()
        if (k == n) {
            result.add(mutableList)
            return
        }
        for (i in k until n) {
            // Swap
            val temp = mutableList[k]
            mutableList[k] = mutableList[i]
            mutableList[i] = temp
            
            generatePermutations(mutableList, k + 1, result)
            
            // Swap back
            mutableList[i] = mutableList[k]
            mutableList[k] = temp
        }
    }

    private fun twoOpt(
        start: Pair<Double, Double>?,
        points: List<MapSurveyItem>
    ): List<MapSurveyItem> {
        if (points.size <= 2) return points
        
        var bestRoute = points.toMutableList()
        var bestDist = calculateTotalDistance(start, bestRoute)
        var improved = true
        var iteration = 0
        val maxIterations = 200
        
        while (improved && iteration < maxIterations) {
            improved = false
            for (i in 0 until bestRoute.size - 1) {
                for (j in i + 1 until bestRoute.size) {
                    val newRoute = twoOptSwap(bestRoute, i, j)
                    val newDist = calculateTotalDistance(start, newRoute)
                    if (newDist < bestDist) {
                        bestRoute = newRoute.toMutableList()
                        bestDist = newDist
                        improved = true
                    }
                }
            }
            iteration++
        }
        return bestRoute
    }

    private fun twoOptSwap(route: List<MapSurveyItem>, i: Int, j: Int): List<MapSurveyItem> {
        val result = mutableListOf<MapSurveyItem>()
        for (c in 0 until i) {
            result.add(route[c])
        }
        for (c in j downTo i) {
            result.add(route[c])
        }
        for (c in j + 1 until route.size) {
            result.add(route[c])
        }
        return result
    }

    private fun calculateTotalDistance(start: Pair<Double, Double>?, route: List<MapSurveyItem>): Double {
        var dist = 0.0
        var prevLat = start?.first
        var prevLng = start?.second
        for (pt in route) {
            if (prevLat != null && prevLng != null) {
                dist += com.example.util.GeoUtils.haversineKm(prevLat, prevLng, pt.latitude, pt.longitude)
            }
            prevLat = pt.latitude
            prevLng = pt.longitude
        }
        return dist
    }
}

File: src/core/engine/route-optimizer.ts
import { calculateDistanceKm } from "../utils/coordinates";

export interface RoutePoint {
  id: string;
  latitude: number;
  longitude: number;
  title: string;
}

export interface OptimizationResult<T extends RoutePoint> {
  optimizedPoints: T[];
  invalidPointsCount: number;
}

function isValidCoordinate(lat: number, lng: number): boolean {
  return lat !== 0 && lng !== 0 && !isNaN(lat) && !isNaN(lng) && lat >= -90 && lat <= 90 && lng >= -180 && lng <= 180;
}

export class RouteOptimizer {
  public static optimize<T extends RoutePoint>(
    start: [number, number] | null,
    points: T[]
  ): OptimizationResult<T> {
    const validPoints = points.filter((pt) => isValidCoordinate(pt.latitude, pt.longitude));
    const invalidCount = points.length - validPoints.length;

    if (validPoints.length === 0) {
      return { optimizedPoints: [], invalidPointsCount: invalidCount };
    }

    const optimized =
      validPoints.length <= 7
        ? this.permute(start, validPoints)
        : this.twoOpt(start, validPoints);

    return {
      optimizedPoints: optimized,
      invalidPointsCount: invalidCount
    };
  }

  private static permute<T extends RoutePoint>(
    start: [number, number] | null,
    points: T[]
  ): T[] {
    if (points.length <= 1) return points;

    let bestRoute = points;
    let minDistance = Number.MAX_VALUE;

    const indices = points.map((_, i) => i);
    const permutations: number[][] = [];
    this.generatePermutations(indices, 0, permutations);

    for (const perm of permutations) {
      let currentDist = 0;
      let prevLat = start ? start[0] : null;
      let prevLng = start ? start[1] : null;

      for (const idx of perm) {
        const pt = points[idx];
        if (prevLat !== null && prevLng !== null) {
          currentDist += calculateDistanceKm(prevLat, prevLng, pt.latitude, pt.longitude);
        }
        prevLat = pt.latitude;
        prevLng = pt.longitude;
      }

      if (currentDist < minDistance) {
        minDistance = currentDist;
        bestRoute = perm.map((i) => points[i]);
      }
    }

    return bestRoute;
  }

  private static generatePermutations(list: number[], k: number, result: number[][]) {
    const n = list.length;
    const arr = [...list];
    if (k === n) {
      result.push(arr);
      return;
    }
    for (let i = k; i < n; i++) {
      const temp = arr[k];
      arr[k] = arr[i];
      arr[i] = temp;

      this.generatePermutations(arr, k + 1, result);

      const revert = arr[k];
      arr[k] = arr[i];
      arr[i] = revert;
    }
  }

  private static twoOpt<T extends RoutePoint>(
    start: [number, number] | null,
    points: T[]
  ): T[] {
    // Nearest neighbor initial route
    const remaining = [...points];
    const route: T[] = [];
    let curLat = start ? start[0] : remaining[0].latitude;
    let curLng = start ? start[1] : remaining[0].longitude;

    if (!start) {
      route.push(remaining.shift()!);
    }

    while (remaining.length > 0) {
      let nearestIdx = 0;
      let minDist = Number.MAX_VALUE;
      for (let i = 0; i < remaining.length; i++) {
        const d = calculateDistanceKm(curLat, curLng, remaining[i].latitude, remaining[i].longitude);
        if (d < minDist) {
          minDist = d;
          nearestIdx = i;
        }
      }
      const nextPt = remaining.splice(nearestIdx, 1)[0];
      route.push(nextPt);
      curLat = nextPt.latitude;
      curLng = nextPt.longitude;
    }

    // 2-opt optimization loop
    let improved = true;
    let iterations = 0;
    while (improved && iterations < 50) {
      improved = false;
      iterations++;
      for (let i = 0; i < route.length - 1; i++) {
        for (let k = i + 1; k < route.length; k++) {
          const delta = this.calculate2OptDelta(start, route, i, k);
          if (delta < -0.0001) {
            // Reverse segment between i and k
            const segment = route.slice(i, k + 1).reverse();
            route.splice(i, segment.length, ...segment);
            improved = true;
          }
        }
      }
    }

    return route;
  }

  private static calculate2OptDelta<T extends RoutePoint>(
    start: [number, number] | null,
    route: T[],
    i: number,
    k: number
  ): number {
    const prevA = i === 0 ? (start ? { latitude: start[0], longitude: start[1] } : null) : route[i - 1];
    const A = route[i];
    const B = route[k];
    const nextB = k + 1 < route.length ? route[k + 1] : null;

    let oldDist = 0;
    let newDist = 0;

    if (prevA) {
      oldDist += calculateDistanceKm(prevA.latitude, prevA.longitude, A.latitude, A.longitude);
      newDist += calculateDistanceKm(prevA.latitude, prevA.longitude, B.latitude, B.longitude);
    }
    if (nextB) {
      oldDist += calculateDistanceKm(B.latitude, B.longitude, nextB.latitude, nextB.longitude);
      newDist += calculateDistanceKm(A.latitude, A.longitude, nextB.latitude, nextB.longitude);
    }

    return newDist - oldDist;
  }
}
